### PR TITLE
[PUBDEV-4572][PUBDEV-4548] Move initialization of XGBoost into H2O co…

### PR DESCRIPTION
**Pull request by jakubhava**
…re extension

Use RestApiContext

- Use RestApiContext inside registerEndpoint method -> Less static calls, easier to change backend in the future
- Create AlgoAbstractRegister with method registering algos. It was previously in the AbstractRegister where it semantically does not belong
- Remove RegisterSourceRoots from extensions. It's not rest api extension. Instead call it explicitly in the same location where it would be called as the extension. This also allows us to remove the relativeResourcePath argument in register method
- Remove H2O.register method as the rest api should be registered via the restApiContext or RequestServer

Fix formating and change log level to warn

Use singleton instead of static methods

Fix test

remove deprecated method in tests

### CODE REVIEW

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

File: h2o-core/src/main/java/water/H2O.java [hunk 1]
```diff
@@ -775,8 +775,10 @@ public static void configureLogging() {
     // The trick is the output path / file isn't known until the H2O API PORT is chosen,
     // so real logger initialization has to happen somewhat late in the startup lifecycle.
     java.util.Properties p = new java.util.Properties();
+    p.setProperty("log4j.rootLogger", "INFO");
     p.setProperty("log4j.logger.org.reflections.Reflections", "WARN");
     p.setProperty("log4j.logger.org.eclipse.jetty", "WARN");
+    p.setProperty("log4j.logger.ml.dmlc.xgboost4j.java", "OFF");
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/H2O.java]: Not necessary, we do not want to hide messages from XGboost.

**Comment by jakubhava** [comment]
@mmalohlava done 👍

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@mmalohlava so even when I enable the logging, the PR_startup_checks still fail since it checks the console output and fails if there is line logged on different level ( or just without level ) then `INFO`.

The other solution different then disabling the logging is to create custom appender which will change the logging level from INFO to WARN.

However I'm not sure if we really need to have logging from XGBoost subsystem at this stage as we already log the relevant messages on our side.

What about keep it disabled at the initialisation phase but re-enable it when we initialise full logging in the `Log`'s method `setLog4jProperties` ?

**Comment by jakubhava** [comment]
Maybe the simplest and the most transparent solution would be to to simply change PR_startup_check and enable `WARN` messages to be logged so we can se from which places XGBoost couldn't be loaded. What do you think ? @mmalohlava

**Comment by jakubhava** [comment]
@h2o-ops please test

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@mmalohlava XGBoost logging switched to `ERROR`

File: h2o-core/src/main/java/water/H2O.java [hunk 2]
```diff
@@ -1627,12 +1534,6 @@ private static void startNetworkServices() {
 
   }
 
-  static public void register(
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/H2O.java]: We need to keep this one for backward compatibility.

**Comment by mmalohlava** [comment]
I think we can merge after small compatibility change ^^ - i would just mark the mentioned method deprecated instead of removing it.

**Comment by jakubhava** [comment]
@mmalohlava done

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Pull request merged**